hw/top_earlgrey/data/placement.xdc

```diff
@@ -2,13 +2,19 @@
 # See here (https://www.xilinx.com/support/answers/66386.html)
 # See also this link (https://github.com/lowRISC/opentitan/pull/8138#issuecomment-916696830)
 # for a thorough explanation of why such a custom placement helps.
-# Note the placement location was not chosen through any systematic means, but through trial
-# and error, it may become necessary in the future to tweak this if other congestion
-# issues arise.
+#
+# The placement location has been chosen by evaluating the resulting place and route times of
+# all possible options.
+#
+# The most suitable locations include: X0Y4, X0Y2, X1Y6, X1Y5
+# The following locations lead to excessive implementation times: X0Y0, X0Y6, X1Y1
+#
+# The evaluation has been performed using commit 85910ce3c2dfd94f860f4002b4be0cd0985aa058.
+# It may become necessary in the future to tweak this if other congestion issues arise.
 
-# Clock net "top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]" driven by instance "top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/gen_gate.gen_bufhce.u_bufhce" located at site "BUFHCE_X1Y0"
+# Clock net "top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]" driven by instance "top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/gen_gate.gen_bufhce.u_bufhce" located at site "BUFHCE_X0Y4"
 #startgroup
 create_pblock {CLKAG_top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]}
 add_cells_to_pblock [get_pblocks  {CLKAG_top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]}] [get_cells -filter { PRIMITIVE_GROUP != I/O && IS_PRIMITIVE==1 && PRIMITIVE_LEVEL !=INTERNAL } -of_object [get_pins -filter {DIRECTION==IN} -of_objects [get_nets -hierarchical -filter {PARENT=="top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]"}]]]
-resize_pblock [get_pblocks {CLKAG_top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]}] -add {CLOCKREGION_X1Y0:CLOCKREGION_X1Y0}
+resize_pblock [get_pblocks {CLKAG_top_earlgrey/u_clkmgr_aon/u_clk_main_aes_cg/gen_xilinx.u_impl_xilinx/clocks_o[clk_main_aes]}] -add {CLOCKREGION_X0Y4:CLOCKREGION_X0Y4}
```

Comment:
This looks like a good spot to me. The only I/Os there are GPIOs, which can have the most liberal timing, so there likely won't be other timing-critical IPs that need to route through. I suspect the reason this is better than being totally out of the way at the top of the device is because AES wants to be close to all the other devices on the high-speed TL-UL crossbar.

Well, technically, for I/Os, there is also the USB stuff headed for the 34-pin user header, but I think that's something we can drop, really. It's awful that we have the usbdev IP need to connect to I/O pins on opposite corners of the device... We could've at least chosen the PMOD in bank 13 :laughing: But that'd still be on opposite sides, so it's not the most comfortable for meeting timing for USB.

Reply:
does that mean we should try to place the `usbdev` module closer to the I/O's?
The `usbdev`, unlike the aes, doesn't actually need to be sync with the high speed portions of the fabric, so I imagine it has some leeway in moving farther away.

Reply:
I'm not sure of your meaning, but `usbdev` isn't targeted by this constraint.

In general, though, I/O IPs should be close to their pads on the die. `usbdev` has set_input_delay and set_output_delay constraints to encourage Vivado to do so for it (for the on-board PHY only--The alternate interface is left as unconstrained). Before this PR, AES targeted the clock region where the on-board PHY pins are, so there was some minor fighting for routing resources.

Area constraints can force the primitives of a group to go somewhere, but it hasn't been needed for any other blocks so far. We'll probably go through another round of this when we add timing constraints for the SPI I/Os. The speed can get high enough there that we shouldn't leave the delays to chance.

Reply:
i think maybe i misunderstood the previous comment. I thought we were saying that the I/O's being far away from usbdev could be a problem? 

and yeah..i think for fpga, it's not likely we'll achieve production speeds with the SPI I/Os...

Reply:
Ah, I was only saying we should probably drop the alternate interface (i.e. the one that has the FPGA trying to go without a transceiver). usbdev's ports have to connect to the alternate interface (bank 16) and the on-board PHY (bank 32).

We don't have I/O timing constraints on the alternate interface, so I would consider it to be luck if it works. But if we _did_ put constraints there, we'd cause usbdev to get pulled in both directions. I suspect that would have deleterious effects... ;)

On the SPI front, that'll be an interesting one. An FPGA ought to be able to handle a little 30 MHz SPI interface, but the pass-through delays are probably going to be tougher, since that's also going from one side of the die to the other (host on bank 33, device on bank 12... that's apparently hooked up to the SAM3X).

Reply:
yeah, the passthrough timing is the one i'm worried about. Since the pad / on-die delays could be pretty significant.
The direct mailbox type i agree with you "should" be okay.

Thanks for explaining about the alternate interface.

Reply:
Related to `usbdev` we were also thinking to further relax the USB timing constraints to speed up CI implementation time as long as we don't have a test exercising USB on the FPGA. Do you think that idea would be worth exploring?

Reply:
The intentions are laudable there, but I think attempts to relax the constraints would be quite misguided. Remember, the FPGA is not just around for the convenience of the CI: It's the only reasonable platform available for software teams to begin developing their code (at least for a good while). We're probably nearing the time we should start treating it with more care, as something that needs to work for _users_ (both for compute and for I/O).

Unfortunately, USB is the one interface that cannot be slowed down, and removing the constraints risks random failures. Really, I think the opposite approach should be taken for the CI--We ought to focus on getting in tests and closing that gap.

You could try getting rid of the alternate set of pins, though. That may help Vivado avoid attempts to balance the delay between the two banks.

Reply:
Thanks for explaining your view point @a-will . I think this makes sense. We should check to remove the alternate set of pins. Or at least move it to the other PMOD.